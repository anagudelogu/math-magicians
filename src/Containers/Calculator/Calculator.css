.calculator {
    width: 100%;
    min-width: 288px;
    max-width: 400px;
    border-radius: 12px;
    background-color: var(--dark-orange);
    box-shadow: 0 2px 64px rgba(0, 0, 0, 0.2);
  }
  
  .calculator__display {
    padding: 1rem;
    text-align: right;
    background-color: var(--dark-gray);
    color: #fff;
    border-top-left-radius: 12px;
    border-top-right-radius: 12px;
  }
  
  .calculator__value {
    font-size: 2rem;
  }
  
  .calculator__keypad {
    max-height: 100%;
    display: grid;
    border-radius: 12px;
    grid-template-columns: repeat(4, 1fr);
    grid-template-rows: repeat(5, 1fr);
  }
  
  .operations.right {
    grid-area: 1/4/6/5;
  }
  
  .operations.right button {
    width: 100%;
  }
  
  .operations.center {
    grid-area: 1/1/2/4;
    display: grid;
    grid-template-columns: repeat(3, 1fr);
  }
  
  .digits {
    grid-area: 2/1/6/4;
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: auto;
  }
  
  .digits :nth-child(1) {
    grid-area: 4/1/5/3;
    border-bottom-left-radius: 12px;
  }
  