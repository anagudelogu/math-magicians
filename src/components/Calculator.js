import React from 'react';
import calculate from '../js/calculate';
import Button from './Button';

export default class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calc: {
        total: null,
        next: null,
        operation: null,
      },
    };
    this.operationsRight = ['รท', 'x', '-', '+', '='];
    this.operationsCenter = ['AC', '+/-', '%'];
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(buttonName) {
    this.setState((state) => ({
      calc: calculate(state.calc, buttonName),
    }));
  }

  createDigits() {
    this.digits = [];

    for (let i = 0; i < 10; i += 1) {
      this.digits.push(
        <Button
          classN="btn digits__btn"
          value={`${i}`}
          handleClick={this.handleClick}
        />,
      );
    }
    return this.digits;
  }

  createOperations(operations) {
    this.operationsArr = [];
    operations.forEach((operation) => {
      this.operationsArr.push(
        <Button
          classN="btn operation__btn"
          value={operation}
          handleClick={this.handleClick}
        />,
      );
    });
    return this.operationsArr;
  }

  render() {
    const { calc } = this.state;
    const { total, next } = calc;
    return (
      <div className="calculator">
        <div className="calculator__display">
          <span className="calculator__value">
            {' '}
            {next || total || '0'}
          </span>
        </div>

        <div className="calculator__keypad">
          <div className="operations right">
            {this.createOperations(this.operationsRight)}
          </div>

          <div className="operations center">
            {this.createOperations(this.operationsCenter)}
          </div>

          <div className="digits">
            {this.createDigits()}
            <Button
              classN="btn digits__btn"
              value="."
              handleClick={this.handleClick}
            />
          </div>
        </div>
      </div>
    );
  }
}
