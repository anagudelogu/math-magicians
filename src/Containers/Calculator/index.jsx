import React, { useState } from 'react';
import calculate from './logic/calculate';
import Button from '../../components/Button';
import './Calculator.css';
import * as keys from './logic/keys';

const Calculator = () => {
  const [state, setState] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const { total, next, operation } = state;

  const handleClick = (buttonName) => {
    setState((prevState) => calculate(prevState, buttonName));
  };

  const createDigits = (digits) => {
    const digitsBtns = digits.map((digit) => (
      <Button
        key={`${digit}`}
        isOperation={false}
        value={`${digit}`}
        handleClick={handleClick}
      />
    ));
    return digitsBtns;
  };

  const createOperations = (operations) => {
    const operationsArr = operations.map((operation) => (
      <Button
        key={operation}
        isOperation
        value={operation}
        handleClick={handleClick}
      />
    ));
    return operationsArr;
  };

  return (
    <div data-testid="Calculator" className="calculator">
      <div className="calculator__display">
        <span data-testid="value" className="calculator__value">
          {' '}
          {next || operation || total || '0'}
        </span>
      </div>

      <div className="calculator__keypad">
        <div className="operations right">
          {createOperations(keys.OPERATIONS_RIGHT)}
        </div>

        <div className="operations center">
          {createOperations(keys.OPERATIONS_CENTER)}
        </div>

        <div className="digits">{createDigits(keys.DIGITS)}</div>
      </div>
    </div>
  );
};

export default Calculator;
