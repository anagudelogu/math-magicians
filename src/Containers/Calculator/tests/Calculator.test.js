import Calculator from '../index';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Calculator', () => {
  it('Renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('User clicked a Button in calculator', () => {
    render(<Calculator />);
    const button = screen.getByText('7');
    const value = screen.getByTestId('value');

    fireEvent.click(button, {target: {innerText: '7'}});

    expect(value.textContent).toBe(` ${button.textContent}`);
  });

  it('Should do operations', () => {
    render(<Calculator />);
    const value = screen.getByTestId('value');

    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('x'));
    fireEvent.click(screen.getByText('7'));
    fireEvent.click(screen.getByText('='));

    expect(value.textContent).toBe(' 35');
  });
});