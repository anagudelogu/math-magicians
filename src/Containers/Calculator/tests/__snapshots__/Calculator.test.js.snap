// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Renders correctly 1`] = `
<div
  className="calculator"
  data-testid="Calculator"
>
  <div
    className="calculator__display"
  >
    <span
      className="calculator__value"
      data-testid="value"
    >
       
      0
    </span>
  </div>
  <div
    className="calculator__keypad"
  >
    <div
      className="operations right"
    >
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
    <div
      className="operations center"
    >
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="btn operation__btn"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
    </div>
    <div
      className="digits"
    >
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="btn "
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
  </div>
</div>
`;
